pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -0,0 +1,540 @@
+			}
+			return node, nil
+		} else if planChangedForUpdate(nodePlan, node.Spec.NodePlan.Plan) {
+			logrus.Infof("plan [%s] changed for update", node.Name)
+			return uh.updateNodePlan(node, cluster)
+		} else {
+			logrus.Infof("plan [%s] hasn't changed, syncing version nodePlan %v cluster %v", node.Name,
+				node.Spec.NodePlan.Version, cluster.Status.NodeVersion)
+			return uh.updateNodePlanVersion(node, cluster)
+		}
+	}
+
+	if node.Status.AppliedNodeVersion == 0 && workerOnly(node.Status.NodeConfig.Role) {
+		logrus.Infof("waiting for [%s] node-agent to sync, updating nodePlan version %v cluster %v", node.Name, node.Spec.NodePlan.Version, cluster.Status.NodeVersion)
+		return uh.updateNodePlanVersion(node, cluster)
+	}
+
+	return node, nil
+}
+
+func (uh *upgradeHandler) updateNodePlan(node *v3.Node, cluster *v3.Cluster) (*v3.Node, error) {
+	nodePlan, err := uh.getNodePlan(node, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("getNodePlan error for node [%s]: %v", node.Name, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+	np := &v3.NodePlan{
+		Plan:    nodePlan,
+		Version: cluster.Status.NodeVersion,
+	}
+	if node.Spec.NodePlan == nil || node.Spec.NodePlan.AgentCheckInterval == 0 {
+		// default
+		np.AgentCheckInterval = nodeserver.DefaultAgentCheckInterval
+	} else {
+		np.AgentCheckInterval = node.Spec.NodePlan.AgentCheckInterval
+	}
+	nodeCopy.Spec.NodePlan = np
+
+	logrus.Infof("updating node [%s] with plan %v", node.Name, np.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, fmt.Errorf("error updating node [%s] with plan %v", node.Name, err)
+	}
+
+	return updated, err
+}
+
+func (uh *upgradeHandler) updateNodePlanVersion(node *v3.Node, cluster *v3.Cluster) (*v3.Node, error) {
+	if node.Spec.NodePlan.Version == cluster.Status.NodeVersion {
+		return node, nil
+	}
+
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Spec.NodePlan.Version = cluster.Status.NodeVersion
+	logrus.Infof("updating node [%s] with plan version %v", node.Name, nodeCopy.Spec.NodePlan.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, err
+	}
+	return updated, err
+
+}
+
+func (uh *upgradeHandler) getNodePlan(node *v3.Node, cluster *v3.Cluster) (*v3.RKEConfigNodePlan, error) {
+	var (
+		nodePlan *v3.RKEConfigNodePlan
+		err      error
+	)
+	if isNonWorkerOnly(node.Status.NodeConfig.Role) {
+		nodePlan, err = uh.nonWorkerPlan(node, cluster)
+	} else {
+		nodePlan, err = uh.workerPlan(node, cluster)
+	}
+	return nodePlan, err
+}
+
+func lock(name string) {
+	clusterLockMutex.Lock()
+	if _, ok := clusterLock[name]; !ok {
+		clusterLock[name] = &sync.RWMutex{}
+	}
+	cMutex := clusterLock[name]
+	clusterLockMutex.Unlock()
+
+	cMutex.Lock()
+}
+
+func unlock(name string) {
+	time.Sleep(5 * time.Second)
+	clusterLockMutex.RLock()
+	cMutex := clusterLock[name]
+	clusterLockMutex.RUnlock()
+
+	cMutex.Unlock()
+}
+
+func (uh *upgradeHandler) upgradeCluster(cluster *v3.Cluster, nodeName string) error {
```

Comment:
do we need to add check for cluster driver so we can return for non-rke?

Reply:
checking here, think this should suffice? https://github.com/rancher/rancher/pull/25451/files#diff-a16c43ed026936b620e2fdb9ca6196b8R74